### PR TITLE
update jax.tree.map to comply with jax 0.4.34

### LINKED ISSUE #1093: Jax 0.4.34 raises an error for None as tree-prefix
Hello:)

[The most recent release of jax 0.4.34](https://github.com/jax-ml/jax/releases/tag/jax-v0.4.34) does no longer support anything of the kind `jax.tree.map(f, None, non-None)`. 

This also affects optax, for example, in [add_decayed_weights](https://github.com/google-deepmind/optax/blob/main/optax/transforms/_adding.py#L50). 

The error can be simply prevented by changing the line to 

```
    updates = jtu.tree_map(
        lambda g, p: None if g is None else g + weight_decay * p, updates, params, is_leaf=lambda x: x is None)
```

I'm happy to take a look and see if there are any other parts affected and provide a pull request if desired. 

Best, 
Alex

### PR DESCRIPTION
**Pull request by a1302z**
Pull Request to fix #1093

### CODE REVIEW

**Review by vroulet — COMMENTED** [review]
Thank you for doing that!
Would it be possible to add a generic test for such functions?
(tree_update_moment, tree_update_infinity_moment and tree_update_moment_per_elem_norm follow the same fix).
Thanks again!

**Comment by a1302z** [comment]
Happy to do that. How would you imagine this test to be fully generic? A simple test case would be something like this (extended to all corresponding functions):
```
  def test_none_arguments(self):
    m = jnp.logspace(-10, 10, num=21, dtype=jnp.bfloat16)  # 1e-10 ... 1e10
    tu.tree_update_moment(None, m, 0.9, 1)
```
Let me know what you think.

**Comment by a1302z** [comment]
I have added test cases for `tree_add_scalar_mul`, `tree_update_moment`, `tree_update_infinity_moment`, `tree_update_moment_per_elem_norm`, `apply_updates`, `incremental_update`, `keep_params_nonnegative`, and `add_decayed_weights`. I have omitted `scale_by_adam` as mu would need to be None, which is not an input argument to the function.

File: optax/_src/update_test.py [hunk 1]
```diff
@@ -79,5 +79,9 @@ def test_periodic_update(self):
           params_2, new_params, atol=1e-10, rtol=1e-5)
 
 
+  def test_none_argument(self):
```
Comment (vroulet) [inline optax/_src/update_test.py:82]: Make it a parameterized test, that is, add `from absl.testing import parameterized` at the top and then write the test as
```
  @parameterized.named_parameters(
      dict(testcase_name='apply_updates', operation=update.apply_updates),
      dict(testcase_name='incremental_update', operation=update.incremental_update),
  )
    def test_none_argument(self, operation)
      x = jnp.array([1., 2., 3.])
      operation(None, x)
  ```

File: optax/transforms/_adding_test.py [hunk 2]
```diff
@@ -91,6 +91,14 @@ def test_add_noise_has_correct_variance_scaling(self):
 
       chex.assert_trees_all_close(updates_i, updates_i_rescaled, rtol=1e-4)
 
+def test_none_argument():
```
Comment (vroulet) [inline optax/transforms/_adding_test.py:94]: Add an indent (the test must be part of the class AddingTest) and add a self argument to the function.

File: optax/transforms/_constraining_test.py [hunk 3]
```diff
@@ -114,6 +114,15 @@ def test_zero_nans(self):
             (jnp.array(False), jnp.array(False), jnp.array(False))))
     chex.assert_trees_all_close(updates, grads)
 
+  def check_none_arguments(self):
```
Comment (vroulet) [inline optax/transforms/_constraining_test.py:117]: call it `test_none_arguments`

Comment (vroulet) [reply optax/transforms/_constraining_test.py:117]: Also why are you testing for this to raise an error?

File: optax/tree_utils/_tree_math_test.py [hunk 4]
```diff
@@ -236,6 +236,13 @@ def test_empty_tree_reduce(self):
       self.assertEqual(tu.tree_sum(tree), 0)
       self.assertEqual(tu.tree_vdot(tree, tree), 0)
 
+  def test_none_arguments(self):
```
Comment (vroulet) [inline optax/tree_utils/_tree_math_test.py:239]: Make it a parameterized test with the different functions

**Review by vroulet — COMMENTED** [review]
Thank you very much, I left you some additional comments

File: optax/transforms/_constraining_test.py (hunk 3 above)
Comment (a1302z) [reply optax/transforms/_constraining_test.py:117]: Done. In this function is a check that raises a ValueError if None is passed as params. Hence, the tree error can only occur if this check is, for some reason, disabled. If we rely on this check not being modified, we can also remove this test case entirely.

File: optax/_src/update_test.py (hunk 1 above)
Comment (a1302z) [reply optax/_src/update_test.py:82]: Done. However, the functions take a different number of arguments. To make it compatible I've used a lambda expression. Also below for tree_math_test. Let me know if this is fine or if there is a better way to deal with that.

Comment (vroulet) [reply optax/_src/update_test.py]: That's perfect, thank you!

File: optax/transforms/_constraining_test.py (hunk 3 above)
Comment (vroulet) [reply optax/transforms/_constraining_test.py]: Right, perfect, thank you for the explanation!

**Review by vroulet — APPROVED** [review]
That's perfect, thank you again @a1302z !

**Pull request merged**